static/app/stores/guideStore.tsx

```diff
@@ -147,44 +157,46 @@ const storeConfig: GuideStoreDefinition = {
       return;
     }
 
-    const guidesContent: GuidesContent = getGuidesContent(this.state.orgSlug);
+    const guidesContent = getGuidesContent(this.state.orgSlug);
+
     // map server guide state (i.e. seen status) with guide content
-    const guides = guidesContent.reduce((acc: Guide[], content) => {
+    const guides = guidesContent.reduce<Guide[]>((acc, content) => {
       const serverGuide = data.find(guide => guide.guide === content.guide);
-      serverGuide &&
-        acc.push({
-          ...content,
-          ...serverGuide,
-        });
+      if (serverGuide) {
+        acc.push({...content, ...serverGuide});
+      }
       return acc;
     }, []);
 
-    this.state.guides = guides;
+    this.state = {...this.state, guides};
     this.updateCurrentGuide();
   },
 
   onCloseGuide(dismissed?: boolean) {
     const {currentGuide, guides} = this.state;
-    // update the current guide seen to true or all guides
-    // if markOthersAsSeen is true and the user is dismissing
-    guides
-      .filter(
-        guide =>
-          guide.guide === currentGuide?.guide ||
-          (currentGuide?.markOthersAsSeen && dismissed)
-      )
-      .forEach(guide => (guide.seen = true));
-    this.state.forceShow = false;
+
+    // update the current guide seen to true or all guides if markOthersAsSeen
+    // is true and the user is dismissing
+    const updatedGuides = guides.map(guide => {
+      const shouldUpdateSeen =
+        guide.guide === currentGuide?.guide ||
+        (currentGuide?.markOthersAsSeen && dismissed);
+
+      return shouldUpdateSeen ? {...guide, seen: true} : guide;
+    });
+
+    this.state = {...this.state, guides: updatedGuides, forceShow: false};
     this.updateCurrentGuide();
   },
 
   onNextStep() {
-    this.state.currentStep += 1;
+    const {currentStep} = this.state;
+    this.state = {...this.state, currentStep: currentStep + 1};
     this.trigger(this.state);
   },
 
   onToStep(step: number) {
-    this.state.currentStep = step;
+    this.state = {...this.state, currentStep: step};
     this.trigger(this.state);
```

Comment:
Any reason for treating this immutable in this PR? It can have implications on any effects listening to the entire state.

Reply:
Because the useLegacyHook stores the entire result of `getState`, and if the result of `getState` doesn't change identity no re-render will happen.

Reply:
Oh, so it's required! ok thats nice to know. I wasnt sure because I thought that it'll just fire a trigger and then the store sets the new state. I welcome the move :D